src/main/java/org/opensearch/jobscheduler/sweeper/JobSweeper.java

```diff
@@ -186,6 +197,82 @@ public void beforeClose() {
         this.fullSweepExecutor.shutdown();
     }
 
+    @Override
+    public Engine.Index preIndex(ShardId shardId, Engine.Index operation) {
+        if (JobSchedulerPlugin.GuiceHolder.getIndicesService() != null
+            && JobSchedulerPlugin.GuiceHolder.getIdentityService() != null
+            && JobSchedulerPlugin.GuiceHolder.getIdentityService().getScheduledJobIdentityManager() != null) {
+            MapperService mapperService = JobSchedulerPlugin.GuiceHolder.getIndicesService()
+                .indexService(shardId.getIndex())
+                .mapperService();
+            ParsedDocument parsedDoc = operation.parsedDoc();
+
+            try {
+                XContentParser parser = JsonXContent.jsonXContent.createParser(
+                    xContentRegistry,
+                    LoggingDeprecationHandler.INSTANCE,
+                    BytesReference.toBytes(parsedDoc.source())
+                );
+                XContentParser.Token token;
+                XContentBuilder operatorBuilder = XContentFactory.contentBuilder(parser.contentType());
+                XContentBuilder builder = XContentFactory.contentBuilder(parser.contentType());
+                builder.startObject();
+                operatorBuilder.startObject();
+                // the start of the parser
+                if (parser.currentToken() == null) {
+                    parser.nextToken();
+                }
+                String currentFieldName = null;
+                while ((token = parser.currentToken()) != null) {
+                    String tokenName = parser.currentName();
+                    if (token == XContentParser.Token.FIELD_NAME) {
+                        currentFieldName = tokenName;
+                        if (OPERATOR.equals(currentFieldName)) {
+                            operatorBuilder.field(currentFieldName);
+                            parser.nextToken();
+                            operatorBuilder.copyCurrentStructure(parser);
+                        } else {
+                            builder.field(currentFieldName);
+                            parser.nextToken();
+                            builder.copyCurrentStructure(parser);
+                        }
+
+                    } else {
+                        parser.nextToken();
+                    }
+                }
+                builder.endObject();
+                operatorBuilder.endObject();
+
+                SourceToParse toParse = new SourceToParse(
+                    shardId.getIndexName(),
+                    operation.id(),
+                    BytesReference.bytes(builder),
+                    XContentType.JSON
+                );
+
+                ParsedDocument docMinusOperator = mapperService.documentMapper().parse(toParse);
```

Comment:
Maybe I'm having trouble reading this, I'm concerned about this, from this code the implication is that the operator information, including the token are accessible by the plugin, since the data is stored within the existing job index.

Reply:
I'm glad you pointed this out because this is where I want the most feedback. 

1. For plugins, plugins will stash their threadcontext before writing to their jobs index so at the time the job details are indexed the user info is not available in the threadcontext.
2. For extensions, extensions will use their service account token when writing to their jobs index so the identity of the requester (service account) is different from the identity of what to associate for the scheduled job.

This is where this convention comes into play. In this model, when writing the job details document to the plugin/extension's job index they would include a field called `operator` to tell the identity system the identity to store alongside the job details. This parses that information from the document. Using the `preIndex` hook of the `IndexOperatorListener` this **modifies** the document before its stored in the plugin/extensions job index and removes it from the document. Instead, that information would be stored in the centralized scheduled job identity index owned by the security plugin.

See this comment in the companion Core PR for more details: https://github.com/opensearch-project/OpenSearch/pull/7573#issuecomment-1574045477